src/rpp/rpp/operators/buffer.hpp

```diff
@@ -0,0 +1,113 @@
+//                    TC Wang 2022 - present.
+// Distributed under the Boost Software License, Version 1.0.
+//    (See accompanying file LICENSE_1_0.txt or copy at
+//          https://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/victimsnino/ReactivePlusPlus
+//
+
+#pragma once
+
+#include <algorithm>
+#include <mutex>
+
+#include <rpp/observables/constraints.hpp>
+#include <rpp/operators/fwd/buffer.hpp>
+#include <rpp/subscribers/constraints.hpp>
+#include <rpp/sources/create.hpp>
+
+IMPLEMENTATION_FILE(buffer_tag);
+
+namespace rpp::details
+{
+
+/// A non-copyable class that provides a copyable on_next for the subscriber and
+/// allows copies of on_next(s) to share the same states.
+template<constraint::decayed_type UpstreamType, constraint::decayed_type DownstreamType>
+struct buffer_on_next : public std::enable_shared_from_this<buffer_on_next<UpstreamType, DownstreamType>>
+{
+    using DownstreamValueType = UpstreamType;
+
+    /// \param count Number of items being bundled. Note when count == 0, we'll
+    /// treat the behavior like when count == 1.
+    explicit buffer_on_next(size_t count) {
+        buckets_.reserve(std::max<size_t>(1ul, count));
+    };
+
+    // Copy & move constructors
+    buffer_on_next(const buffer_on_next& other) = delete;
+    buffer_on_next(buffer_on_next&&) noexcept = default;
+
+    // Copy & move assignment operators
+    buffer_on_next& operator=(const buffer_on_next&) = delete;
+    buffer_on_next& operator=(buffer_on_next&&) noexcept = default;
+
+    /// \return a on_next function for subscriber, that is copyable and all the
+    /// copies share the same state.
+    auto get_on_next() {
+        return [shared_this = this->shared_from_this()](
+                const UpstreamType& value,
+                const constraint::subscriber_of_type<DownstreamType> auto& subscriber)
+        {
+            std::lock_guard<std::mutex> guard(shared_this->mutex_);
+
+            shared_this->buckets_.push_back(value);
+            if (shared_this->buckets_.size() == shared_this->buckets_.capacity())
+            {
+                auto copy = std::vector<DownstreamValueType>(shared_this->buckets_);
+
+                shared_this->buckets_.clear();
+                subscriber.on_next(std::move(copy));
+            }
+        };
+    }
+
+    /// \return a on_completed function for subscriber, that is copyable and all the
+    /// copies share the same state.
+    auto get_on_completed() {
+        return [shared_this = this->shared_from_this()](
+                const constraint::subscriber_of_type<DownstreamType> auto& subscriber)
+        {
+            std::lock_guard<std::mutex> guard(shared_this->mutex_);
+
+            if (!shared_this->buckets_.empty())
+            {
+                auto copy = std::vector<DownstreamValueType>(shared_this->buckets_);
+                shared_this->buckets_.clear();
+                subscriber.on_next(std::move(copy));
```

Comment:
You can avoid extra copy
```suggestion
                subscriber.on_next(std::move(shared_this->buckets_));
                shared_this->buckets_.clear();
```

Reply:
Hmm, what if the `on_next()` in the middle of long chain in downstream raises an error. But then there's an `retry` operator appending after the downstream subscriber that raises the error. This observable gets resubscribed and the state here is not reset.

For the above reason, I thought it'd be better to reset the inner state and then propagate events to the downstream.

What do you think?

Reply:
Ok, let's check 2 situations:

1) (observable)-....-(buffer)-....-(retry)-....
in this case if any error between buffer and retry, than whole chain between observable and retry would be resubscribed and whole internal state recreated. For you it means that whole `buffer_on_next` will be recreated -> no any problems

2) (observable)-....-(retry)-.....-(buffer)-....
in this case if any error between observable and retry, than whole chain between observable and retry would be resubscribed, but buffer doesn't know about it  -> no any issues

Reply:
Thanks for clarification!

BTW, I think it's a typo in your code suggestion:

```
subscriber.on_next(std::move(shared_this->buckets_));
```

And this is what I think the right version:

```
subscriber.on_next(shared_this->buckets_);
```

We may still have a copy after `on_next()...` chain.

Reply:
Oh my bad, I just found that the `buckets_` will be cleared out after `std::move()`.

Reply:
Ohhhh, it makes sense... Then yes, it is better to have some variable to store max size OR extract size before move and then make re-reserve